generator/camera_info_collector.cpp

```diff
@@ -0,0 +1,317 @@
+    }
+
+    if (curMinDist < minDist && curMinDist < kMaxDistFromCameraToClosestSegmentMeters)
+    {
+      minDist = curMinDist;
+      bestFeatureId = ft.GetID().m_index;
+      segmentIdOfBestFeatureId = static_cast<uint32_t>(bestSegment.m_ind);
+      coef = tmpCoef;
+    }
+  };
+
+  dataSource.ForEachInRect(
+      updClosestFeatureCallback,
+      MercatorBounds::RectByCenterXYAndSizeInMeters(m_center, kSearchCameraRadiusMeters),
+      scales::GetUpperScale());
+
+  if (minDist != std::numeric_limits<double>::max())
+    m_ways.emplace_back(bestFeatureId, segmentIdOfBestFeatureId, coef);
+}
+
+boost::optional<uint32_t> CamerasInfoCollector::Camera::FindMyself(
+    uint32_t wayId, FrozenDataSource const & dataSource, MwmSet::MwmId const & mwmId)
+{
+  FeatureID ft(mwmId, wayId);
+  uint32_t result = 0;
+  bool isRoad = true;
+
+  auto const readFeature = [&result, this, &isRoad, wayId](FeatureType & ft)
+  {
+    bool found = false;
+    isRoad = routing::IsRoad(feature::TypesHolder(ft));
+    if (!isRoad)
+      return;
+
+    auto const forEachPoint = [&result, &found, this](m2::PointD const & pt)
+    {
+      if (found)
+        return;
+
+      if (AlmostEqualAbs(m_center, pt, kEqualityEps))
+        found = true;
+
+      result++;
+    };
+
+    ft.ForEachPoint(forEachPoint, scales::GetUpperScale());
+
+    CHECK(found, ("Cannot find camera point at the feature:", wayId));
+  };
+
+  dataSource.ReadFeature(readFeature, ft);
+
+  return isRoad ? boost::optional<uint32_t>(result) : boost::optional<uint32_t>();
+}
+
+void CamerasInfoCollector::Camera::TranslateWaysIdFromOsmId(std::map<osm::Id, uint32_t> const & osmIdToFeatureId)
+{
+  for (auto it = m_ways.begin(); it != m_ways.end();)
+  {
+    auto const mapIt = osmIdToFeatureId.find(osm::Id::Way(it->m_featureId));
+    if (mapIt == osmIdToFeatureId.cend())
+    {
+      // It means, that way was not valid, and didn't pass to mwm.
+      // So we should also erase it from our set.
+      m_ways.erase(it);
+    }
+    else
+    {
+      it->m_featureId = mapIt->second;  // osmId -> featureId
+      ++it;
+    }
+  }
+}
+
+void CamerasInfoCollector::Camera::Serialize(FileWriter & writer, CamerasInfoCollector::Camera const & camera, uint32_t prevFeatureId)
+{
+  auto waysNumber = base::asserted_cast<uint8_t>(camera.m_ways.size());
+  WriteVarUint(writer, waysNumber);
+
+  for (auto const & way : camera.m_ways)
+  {
+    auto featureId = static_cast<uint32_t>(way.m_featureId);
+    featureId -= prevFeatureId; // delta coding
+    WriteVarUint(writer, featureId);
+    WriteVarUint(writer, way.m_segmentId);
+
+    auto const coef = static_cast<uint32_t>(way.k * kSignificantPartOfSegmentCoef);
+    WriteVarUint(writer, coef);
+  }
+
+  auto const speed = static_cast<uint8_t>(camera.m_maxSpeed);
+  WriteVarUint(writer, speed);
+
+  auto const direction = static_cast<uint8_t>(camera.m_direction);
+  WriteVarUint(writer, direction);
+
+  // TODO add implementation of this feature
```

Comment:
It's better to write which feature should be implemented.

Reply:
done